supply/spec/client_spec.rb

```diff
@@ -8,23 +8,47 @@
         to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
     end
 
-    it "displays error messages from the API" do
-      stub_request(:post, "https://androidpublisher.googleapis.com/upload/androidpublisher/v3/applications/test-app/edits/1/listings/en-US/icon").
-        to_return(status: 403, body: '{"error":{"message":"Ensure project settings are enabled."}}', headers: { 'Content-Type' => 'application/json' })
+    describe "displays error messages from the API" do
+      it "with no retries" do
+        stub_request(:post, "https://androidpublisher.googleapis.com/upload/androidpublisher/v3/applications/test-app/edits/1/listings/en-US/icon").
+          to_return(status: 403, body: '{"error":{"message":"Ensure project settings are enabled."}}', headers: { 'Content-Type' => 'application/json' })
 
-      expect(UI).to receive(:user_error!).with("Google Api Error: Invalid request - Ensure project settings are enabled.")
+        expect(UI).to receive(:user_error!).with("Google Api Error: Invalid request - Ensure project settings are enabled.").once
 
-      current_edit = double
-      allow(current_edit).to receive(:id).and_return(1)
+        current_edit = double
+        allow(current_edit).to receive(:id).and_return(1)
 
-      client = Supply::Client.new(service_account_json: StringIO.new(service_account_file), params: { timeout: 1 })
-      allow(client).to receive(:ensure_active_edit!)
-      allow(client).to receive(:current_edit).and_return(current_edit)
+        client = Supply::Client.new(service_account_json: StringIO.new(service_account_file), params: { timeout: 1 })
+        allow(client).to receive(:ensure_active_edit!)
+        allow(client).to receive(:current_edit).and_return(current_edit)
 
-      client.begin_edit(package_name: 'test-app')
-      client.upload_image(image_path: fixture_file("playstore-icon.png"),
-                          image_type: "icon",
-                            language: "en-US")
+        client.begin_edit(package_name: 'test-app')
+        client.upload_image(image_path: fixture_file("playstore-icon.png"),
+                            image_type: "icon",
+                              language: "en-US")
+      end
+
+      it "with 5 retries" do
+        stub_const("ENV", { 'SUPPLY_UPLOAD_MAX_RETRIES' => 5 })
+
+        stub_request(:post, "https://androidpublisher.googleapis.com/upload/androidpublisher/v3/applications/test-app/edits/1/listings/en-US/icon").
+          to_return(status: 403, body: '{"error":{"message":"Ensure project settings are enabled."}}', headers: { 'Content-Type' => 'application/json' })
+
+        expect(UI).to receive(:error).with("Google Api Error: Invalid request - Ensure project settings are enabled. - Retrying...").exactly(5).times
+        expect(UI).to receive(:user_error!).with("Google Api Error: Invalid request - Ensure project settings are enabled.").once
```

Comment:
One thing that I got bitten by in the past is that when you stub `UI.user_error!` with `expect` or `allow`, that means that the original method won't be called… which means that it won't `raise`.

All that means that the caller of `call_google_api`, in your case `client.upload_image`, won't be interrupted by the `Fastlane::Error` raised by `user_error!` in the context of running tests, while it will in the real world. Depending on the implementation of `client.upload_image(…)`, that could have side effects if that implementation counts on the fact that `call_google_api` can raise to stop early and not run the rest of the code of that method on error, as that would mean that the fact that we stubbed `user_error!` would make the method under test behave differently in test environment vs in reality.

I think the easiest way to solve this would be to replace your `expect(UI).to receive(:user_error!)` with wrapping your call to `client.upload_image` within `expect { … }.to raise_error(Fastlane::Error, "Google Api Error: Invalid request – Ensure project settings are enabled."`. That way you'd let `user_error!` raise and interrupt the code of `upload_image` mid-air even in the context of testing.

Reply:
@AliSoftware That expect on `UI.user_error!` was already there 🤷 But... would be adding `.and_call_original` work? 🤔

Reply:
Well that would, except that I _think_ `rubocop` might suggest you to only call `.and_call_original` on `allow(…)`, not on `expect(…)`

So you could `allow(UI).to receive(:user_error!).with(…).and_call_original`… but if you do that and even if you wanted to still keep the `expect(UI).to receive(:user_error!).once` to assert on it only being called once and not more, you'll still need to wrap your `client.upload_image` within `expect { … }.to raise_error(…)` to avoid the test from crashing on the raised error and instead catch (and assert) it.

So at this point since you'll still need the `expect { … } .to raise_error`, I'm not sure that `and_call_original` adds much value, instead of not stubbing `UI.user_error!` at all?